File: javatest/src/main/java/com/example/javatest/multithread/BlockQueueTest.java
package com.example.javatest.multithread;

import java.util.concurrent.ArrayBlockingQueue;

/**
 * @Author: luweiming
 * @Description:
 * @Date: Created in 15:32 2022/11/9
 */
public class BlockQueueTest {
    private static ArrayBlockingQueue<Integer> queue = new ArrayBlockingQueue<>(10, true);

    public static void main(String[] args) {
//        ThreadGroup group1 = new ThreadGroup("producer");
//        ThreadGroup group2 = new ThreadGroup("consumer");
        new Thread(new Producer(), "producer").start();
        new Thread(new Consumer(), "consumer").start();
    }

    static class Producer implements Runnable {
        @Override
        public void run() {
            try {
                for (int i = 0; i < 20; i++) {
                    Thread.sleep(500);
                    queue.put(10);
                    System.out.println(Thread.currentThread().getName() + ":put");
                }
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
    }

    static class Consumer implements Runnable {
        @Override
        public void run() {
            try {
                for (int i = 0; i < 20; i++) {
                    Thread.sleep(1000);
                    queue.take();
                    System.out.println(Thread.currentThread().getName() + ":take");
                }
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
    }
}

File: app/src/main/java/com/example/androidlearn/multithread/BlockQueueTest.kt
package com.example.androidlearn.multithread

import java.util.concurrent.LinkedBlockingQueue
import java.util.concurrent.TimeUnit
import kotlin.concurrent.thread

/**
 * @Author: luweiming
 * @Description:
 * @Date: Created in 10:05 2023/2/22
 */
fun main() {
    val blockQueue = LinkedBlockingQueue<Int>(10)
    thread(name = "Producer") {
        while (true) {
            //offer不阻塞
            val offerResult = blockQueue.offer(1)
            println(Thread.currentThread().name + offerResult)
        }
    }
    thread(name = "Consumer") {
        while (true) {
            //poll不阻塞
            val pollResult = blockQueue.poll(10, TimeUnit.MILLISECONDS)
            println(Thread.currentThread().name + pollResult)

        }
    }
}
